### DISCUSSION TITLE
only ordered task lists should count towards task total

### DISCUSSION THREAD

**Discussion by thescientist13** [community/community#4836] [Projects and Issues]
## Summary
Distinguish between task list types (ordered vs unordered) in an issue to provide users better management of the task count that gets displayed in the GitHub UI.

## Description
Often times an issue (via template or hand made) may contain a list of options to prompt a user through a set of options, like the below to determine what the issue "type" is:
```md
## Issue Type
- [ ] New Feature
- [ ] Documentation
- [ ] Chore
- [ ] Bug
- [ ] Other
```

Assuming a user "checks" N number of those items, the issue total will display as N / 5 tasks in the UI
<img width="283" alt="Screen Shot 2021-07-29 at 5 03 49 PM" src="https://user-images.githubusercontent.com/895923/127566500-622f4d37-577a-4768-8a2a-91efb502fb68.png">

> _(One could argue the above "interface" is intended to mirror that of a radio UI, but i'm not sure how you do that in markdown?)_

Probably even more common in an issue is to (also) track tasks, goals, TODOs, or other issues (like an Epic). 
 Whatever you need to do to "close" this issue.  So you might add something like this.

```md
## Requirements
1. [ ] Implement the `FooClass` that should do XYZ behavior
1. [ ] Write tests for that behavior
1. [ ] Write API docs
```

Combined with the above issue type list, you'll now see N / 8 tasks in the UI.
<img width="332" alt="Screen Shot 2021-07-29 at 5 11 38 PM" src="https://user-images.githubusercontent.com/895923/127566566-c712e4ed-4573-4b29-a0c3-94e5a1a5c301.png">

## Proposal
Given the above example, if an **_ordered task list_** is used in a GitHub issue, only those tasks should count towards the task count.
```
## Issue Type
- [x] New Feature
- [ ] Documentation
- [ ] Chore
- [ ] Bug
- [ ] Other

## Requirements
1. [ ] Implement the `FooClass` that should do XYZ behavior
1. [ ] Write tests for that behavior
1. [ ] Write API docs
```

So in this case you would actually see 3 Tasks in the GitHub UI (since none of the tasks in the **Requirements** section are checked, but there are 3 of them)
<img width="190" alt="Screen Shot 2021-07-29 at 5 22 20 PM" src="https://user-images.githubusercontent.com/895923/127567644-30b7d2b7-a5d6-4e66-a444-a46803ff0a1a.png">

## Thoughts
Some thoughts / questions:
1. Is there a radio button option in markdown?  Could that be used to influence a task count?
1. Or will the concept of issue "type" be part of the new projects feature coming to GitHub?

If both are no, then would love to see this feature!  🙏

**Comment by knowler** [comment]
> Often times an issue (via template or hand made) may contain a list of options to prompt a user through a set of options, like the below to determine what the issue "type" is:

Why not use [multiple issue templates](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/about-issue-and-pull-request-templates) or [labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels) for this? Those are the appropriate GitHub features to use for this.

Forcing the use of ordered lists in Markdown is very opinionated and would force users to use improper semantics (i.e. ordering a list that doesn’t need to be ordered). When I’m making a list of tasks for an issue, I don’t necessarily care what order they are achieved in. When I do care, I’ll use an ordered list.

**Reply by thescientist13** [reply]
> Why not use multiple issue templates or labels for this? Those are the appropriate GitHub features to use for this.

Ah, yeah, I guess that would help the "prompt for more information part" by moving it more to the workflow then the implementation / issue.

I think labels less so because users (at least those not part of the repo already) often don't have permissions to set those, but the template issue types in this case would remove the need to have the first set of tasks at all, which I hadn't thought of.  💡 

> When I’m making a list of tasks for an issue, I don’t necessarily care what order they are achieved in. When I do care, I’ll use an ordered list.

True, that's a fair point as well.

----

Good stuff, thanks for the ideas and suggestions, I hadn't used issue templates much yet but that does seems like a good option for this case.  👍

**Reply by knowler** [reply]
> I think labels less so because users (at least those not part of the repo already) often don't have permissions to set those, but the template issue types in this case would remove the need to have the first set of tasks at all, which I hadn't thought of. 💡

Yep, good point.

Multiple issue templates works best in my experience. GitHub’s has a pretty clean UI for choosing the issue type. You can even add non-issue template actions to the selection screen which is helpful for filtering out support requests that you don’t want showing up in the repo’s issue tracker.

Here’s an example: 

https://github.com/roots/.github/blob/2fcef226452810e4ea84fb9fcfaea892c5409f53/.github/ISSUE_TEMPLATE/config.yml

<img width="914" alt="image" src="https://user-images.githubusercontent.com/6908001/127671269-7b9c6315-b275-46de-83b6-504c447d6eda.png">

You can even disallow using the blank template if I recall correctly.

**Reply by thescientist13** [reply]
Cool, thank you.  Will give that a try!  💯

**Reply by dskvr** [reply]
Multiple issue templates do not address every use case. The example given by @thescientist13 is not the best example to argue for this feature, where you are right to state that Issue Templates are a better solution. 

Better examples are:

```
Please select affected versions:
- [ ] 1.1
- [ ] 1.2
- [ ] 1.3 
```

or...

```
- [ ] This change requires documentation changes
```

or...

```
Please select affected platforms:
- [ ] Windows
- [ ] MacOS
- [ ] Ubuntu
```

[Live example](https://github.com/EOSIO/eos/pull/10581) (1.1k watching, 11.2K stars, 3.6k forks)

Assuming checkboxes are always a task is erroneous and this problem has been brought up many times before on various platforms. **We need a way to distinguish between a task list and just a checkbox.**

The argument that "Forcing the use of ordered lists in Markdown is very opinionated and would force users to use improper semantics" is no more opinionated than turning every checkbox into a task. And if `1. [ ]` is an "improper semantic" in the context of Markdown then so is `- [ ]`, as none of them are included in any Markdown AST outside of Github. But really, there's no such thing as "official markdown syntax", which is why Markdown has `n` flavors and a kind of a chaotic ecosystem to begin with. `asciidoc` users would probably argue that Markdown itself is improper semantics. 

**There is absolutely nothing semantically wrong with this idea in the context of Markdown**

**Reply by knowler** [reply]
@dskvr I respectfully disagree (and frankly I don’t care to give explanation — I don’t have time).

I do agree that this is a problem, and that the multiple issue template and label solutions that I suggested do not solve every situation as you’ve noted. Perhaps as an alternative, GitHub should take some inspiration from [Org mode](https://orgmode.org)’s `TODO` and `DONE` syntax for tasks. It could work like this:

```markdown
- TODO Some task
- DONE Some other task
- TODO Another task

1. DONE Some task to complete _first_.
2. TODO Some task to complete _second_.
2. TODO Some task to complete _third_.

- [ ] Not a task

1. [ ] Not a task
```

Edit: on further reflection, I realize that the syntax direction I am proposing is biased towards English whereas Markdown syntax typically remains language-neutral. I think the direction of introducing new syntax rather than modifying existing syntax (and existing semantics) is still the best direction.

**Reply by thescientist13** [reply]
maybe it would also help if I rephrased the title of this issue?  I certainly didn't want it to come across as a "my way or the highway" proposal given the nature of it breaking backwards compat.  Just wanted to make sure it was stated clearly what was desired is all.

So maybe something like
- Ability to influence task count totals in issues
- Alternatives for managing task count totals in issues

?

**Comment by dskvr** [comment]
I like this idea. Using bullets for one purpose and numbered lists for another is a very elegant and simple solution to this contentious issue. 

The only issue I see with this suggestion is that existing task lists would break, but could be handled with an repo setting that modifies markdown behaviors.

**Reply by thescientist13** [reply]
Yeah, that's why I was also curious if there was a "radio" button option for markdown, or could GitHub introduce one?  That way, for those who want to introduce a one-choice only prompt, they could opt-in to this new feature
```
1. ( ) Yes
1. ( ) No
```


And those would _not_ count towards the task total. Thus with a more complete example like in the description
```
# Some Info
1. ( ) Yes
1. ( ) No

# Some Tasks
1. [ ] Thing 1
1. [ ] Thing 2
```

You would get two task counts on your issue.

Best part, this would be backwards compatible since only the new syntax / markup would apply and thus would be explicitly _opt-in_.  🙌

**Reply by dskvr** [reply]
Github already has their own flavor of markdown that diverges slightly from the other 10+ popular flavors, so adding new markdown syntax would not be difficult to achieve
